In [1]:
'''
This python program scrapes TSwift's "Shake it off" and performs super basic sentiment analysis
using the NLTK library and vader_lexicon

This example demonstrates using the NLTK library, bs4, passing functions as arguments

Jimmy D Nguyen 11-12-2022
'''


from bs4 import BeautifulSoup
import requests
url = 'https://www.azlyrics.com/lyrics/taylorswift/shakeitoff.html'
page = requests.get(url)

In [2]:
page.content

b'<!DOCTYPE html>\r\n<html lang="en">\r\n<head>\r\n<meta charset="utf-8">\r\n<meta http-equiv="X-UA-Compatible" content="IE=edge">\r\n<meta name="viewport" content="width=device-width, initial-scale=1">\r\n<meta name="description" content="Taylor Swift &quot;Shake It Off&quot;: I stay out too late Got nothing in my brain That\'s what people say, mmm-mmm That\'s what people say,..."> \r\n<meta name="keywords" content="Shake It Off lyrics, Taylor Swift Shake It Off lyrics, Taylor Swift lyrics">\r\n<meta name="robots" content="noarchive">\r\n<meta property="og:image" content="//www.azlyrics.com/az_logo_tr.png">\r\n<title>Taylor Swift - Shake It Off Lyrics | AZLyrics.com</title>\r\n\r\n<link rel="canonical" href="https://www.azlyrics.com/lyrics/taylorswift/shakeitoff.html" />\r\n<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/css/bootstrap.min.css">\r\n<link rel="stylesheet" href="/local/az.css">\r\n\r\n<!-- HTML5 shim and Respond.js for IE8 support of HTML5 el

In [3]:
soup = BeautifulSoup(page.content, 'html.parser')

In [4]:
soup

<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1" name="viewport"/>
<meta content="Taylor Swift &quot;Shake It Off&quot;: I stay out too late Got nothing in my brain That's what people say, mmm-mmm That's what people say,..." name="description"/>
<meta content="Shake It Off lyrics, Taylor Swift Shake It Off lyrics, Taylor Swift lyrics" name="keywords"/>
<meta content="noarchive" name="robots"/>
<meta content="//www.azlyrics.com/az_logo_tr.png" property="og:image"/>
<title>Taylor Swift - Shake It Off Lyrics | AZLyrics.com</title>
<link href="https://www.azlyrics.com/lyrics/taylorswift/shakeitoff.html" rel="canonical"/>
<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/css/bootstrap.min.css" rel="stylesheet"/>
<link href="/local/az.css" rel="stylesheet"/>
<!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
<!--[if lt IE 9]>

In [5]:
lyrics = soup.find_all('div', class_='col-xs-12 col-lg-8 text-center')

In [6]:
# len(lyrics) was 1 element only

lyrics = lyrics[0].find_all('div')

In [7]:
lyrics

[<div class="div-share noprint">
 <div class="addthis_toolbox addthis_default_style">
 <a class="btn btn-xs btn-share addthis_button_facebook" href="https://www.facebook.com/sharer.php?u=https%3A%2F%2Fwww.azlyrics.com%2Flyrics%2Ftaylorswift%2Fshakeitoff.html&amp;title=Taylor%20Swift%20-%20Shake%20It%20Off%20Lyrics%20%7C%20AZLyrics.com&amp;display=page" target="_blank">
 <span class="playblk"><img alt="Share on Facebook" class="playblk" height="25" src="//www.azlyrics.com/images/share-icons/facebook.svg" title="Share on Facebook" width="25"/></span>
 </a>
 <a class="btn btn-xs btn-share addthis_button_twitter" href="https://twitter.com/intent/tweet?url=https%3A%2F%2Fwww.azlyrics.com%2Flyrics%2Ftaylorswift%2Fshakeitoff.html&amp;text=Taylor%20Swift%20-%20Shake%20It%20Off%20Lyrics%20%7C%20AZLyrics.com" target="_blank">
 <span class="playblk"><img alt="Share on Twitter" class="playblk" height="25" src="//www.azlyrics.com/images/share-icons/twitter.svg" title="Share on Twitter" width="25"/><

In [8]:
lyrics[5].text

'\n\r\nI stay out too late\nGot nothing in my brain\nThat\'s what people say, mmm-mmm\nThat\'s what people say, mmm-mmm\n\nI go on too many dates\nBut I can\'t make them stay\nAt least that\'s what people say, mmm-mmm\nThat\'s what people say, mmm-mmm\n\nBut I keep cruising\nCan\'t stop, won\'t stop moving\nIt\'s like I got this music\nIn my mind\nSaying, "It\'s gonna be alright."\n\n\'Cause the players gonna play, play, play, play, play\nAnd the haters gonna hate, hate, hate, hate, hate\nBaby, I\'m just gonna shake, shake, shake, shake, shake\nI shake it off, I shake it off\nHeart-breakers gonna break, break, break, break, break\nAnd the fakers gonna fake, fake, fake, fake, fake\nBaby, I\'m just gonna shake, shake, shake, shake, shake\nI shake it off, I shake it off\n\nI never miss a beat\nI\'m lightning on my feet\nAnd that\'s what they don\'t see, mmm-mmm\nThat\'s what they don\'t see, mmm-mmm\n\nI\'m dancing on my own (dancing on my own)\nI make the moves up as I go (moves up as I 

In [9]:
shakeitoff = lyrics[5].text

In [10]:
shakeitoff

'\n\r\nI stay out too late\nGot nothing in my brain\nThat\'s what people say, mmm-mmm\nThat\'s what people say, mmm-mmm\n\nI go on too many dates\nBut I can\'t make them stay\nAt least that\'s what people say, mmm-mmm\nThat\'s what people say, mmm-mmm\n\nBut I keep cruising\nCan\'t stop, won\'t stop moving\nIt\'s like I got this music\nIn my mind\nSaying, "It\'s gonna be alright."\n\n\'Cause the players gonna play, play, play, play, play\nAnd the haters gonna hate, hate, hate, hate, hate\nBaby, I\'m just gonna shake, shake, shake, shake, shake\nI shake it off, I shake it off\nHeart-breakers gonna break, break, break, break, break\nAnd the fakers gonna fake, fake, fake, fake, fake\nBaby, I\'m just gonna shake, shake, shake, shake, shake\nI shake it off, I shake it off\n\nI never miss a beat\nI\'m lightning on my feet\nAnd that\'s what they don\'t see, mmm-mmm\nThat\'s what they don\'t see, mmm-mmm\n\nI\'m dancing on my own (dancing on my own)\nI make the moves up as I go (moves up as I 

In [11]:
shakeitoff_split = shakeitoff.split(sep='\n')

In [12]:
shakeitoff_split

['',
 '\r',
 'I stay out too late',
 'Got nothing in my brain',
 "That's what people say, mmm-mmm",
 "That's what people say, mmm-mmm",
 '',
 'I go on too many dates',
 "But I can't make them stay",
 "At least that's what people say, mmm-mmm",
 "That's what people say, mmm-mmm",
 '',
 'But I keep cruising',
 "Can't stop, won't stop moving",
 "It's like I got this music",
 'In my mind',
 'Saying, "It\'s gonna be alright."',
 '',
 "'Cause the players gonna play, play, play, play, play",
 'And the haters gonna hate, hate, hate, hate, hate',
 "Baby, I'm just gonna shake, shake, shake, shake, shake",
 'I shake it off, I shake it off',
 'Heart-breakers gonna break, break, break, break, break',
 'And the fakers gonna fake, fake, fake, fake, fake',
 "Baby, I'm just gonna shake, shake, shake, shake, shake",
 'I shake it off, I shake it off',
 '',
 'I never miss a beat',
 "I'm lightning on my feet",
 "And that's what they don't see, mmm-mmm",
 "That's what they don't see, mmm-mmm",
 '',
 "I'm da

In [13]:
shakeitoff_processed = []

for i in shakeitoff_split:
    if i != '':
        if i != '\r':
            print(i)
            shakeitoff_processed.append(i)
        
        

I stay out too late
Got nothing in my brain
That's what people say, mmm-mmm
That's what people say, mmm-mmm
I go on too many dates
But I can't make them stay
At least that's what people say, mmm-mmm
That's what people say, mmm-mmm
But I keep cruising
Can't stop, won't stop moving
It's like I got this music
In my mind
Saying, "It's gonna be alright."
'Cause the players gonna play, play, play, play, play
And the haters gonna hate, hate, hate, hate, hate
Baby, I'm just gonna shake, shake, shake, shake, shake
I shake it off, I shake it off
Heart-breakers gonna break, break, break, break, break
And the fakers gonna fake, fake, fake, fake, fake
Baby, I'm just gonna shake, shake, shake, shake, shake
I shake it off, I shake it off
I never miss a beat
I'm lightning on my feet
And that's what they don't see, mmm-mmm
That's what they don't see, mmm-mmm
I'm dancing on my own (dancing on my own)
I make the moves up as I go (moves up as I go)
And that's what they don't know, mmm-mmm
That's what they

In [14]:
shakeitoff_processed

['I stay out too late',
 'Got nothing in my brain',
 "That's what people say, mmm-mmm",
 "That's what people say, mmm-mmm",
 'I go on too many dates',
 "But I can't make them stay",
 "At least that's what people say, mmm-mmm",
 "That's what people say, mmm-mmm",
 'But I keep cruising',
 "Can't stop, won't stop moving",
 "It's like I got this music",
 'In my mind',
 'Saying, "It\'s gonna be alright."',
 "'Cause the players gonna play, play, play, play, play",
 'And the haters gonna hate, hate, hate, hate, hate',
 "Baby, I'm just gonna shake, shake, shake, shake, shake",
 'I shake it off, I shake it off',
 'Heart-breakers gonna break, break, break, break, break',
 'And the fakers gonna fake, fake, fake, fake, fake',
 "Baby, I'm just gonna shake, shake, shake, shake, shake",
 'I shake it off, I shake it off',
 'I never miss a beat',
 "I'm lightning on my feet",
 "And that's what they don't see, mmm-mmm",
 "That's what they don't see, mmm-mmm",
 "I'm dancing on my own (dancing on my own)",

In [15]:
import nltk
from nltk.sentiment import SentimentIntensityAnalyzer

In [16]:
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\jimmy\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [17]:
analyzer = SentimentIntensityAnalyzer()

In [18]:
len(shakeitoff_processed)

74

In [19]:
# Function for sentiment aggregator

def sentiment_calculator(text):
    compound_score = 0
    for i in text:
        score_result = analyzer.polarity_scores(i)
        compound_score = compound_score + score_result['compound']
    return compound_score/float(len(text))

In [20]:
sentiment_calculator(shakeitoff_processed)

-0.19323513513513513

In [39]:
# line by line analysis

# this demonstrates passing a function as an argument
# important to be function=function in order to have a default 
# or the parent function will want an argument pass

def line_analyser(lyrics_list, sentiment_calculator=sentiment_calculator):
    for i in lyrics_list:
        compound_score = analyzer.polarity_scores(i)['compound']
        print(f"{i}: ==> {compound_score}")
    print(f'Aggregate Compound Score: {sentiment_calculator(lyrics_list)}')
        
    

In [36]:
# processing line by line and evaluating the sentiment with -1 being the most negative
# +1 being the most positive and 0 being neutral
# the aggregate score is the very end as the Compound Score

line_analyser(shakeitoff_processed)

I stay out too late: ==> 0.0
Got nothing in my brain: ==> 0.0
That's what people say, mmm-mmm: ==> 0.0
That's what people say, mmm-mmm: ==> 0.0
I go on too many dates: ==> 0.0
But I can't make them stay: ==> 0.0
At least that's what people say, mmm-mmm: ==> 0.0
That's what people say, mmm-mmm: ==> 0.0
But I keep cruising: ==> 0.0
Can't stop, won't stop moving: ==> 0.4168
It's like I got this music: ==> 0.3612
In my mind: ==> 0.0
Saying, "It's gonna be alright.": ==> 0.0
'Cause the players gonna play, play, play, play, play: ==> 0.875
And the haters gonna hate, hate, hate, hate, hate: ==> -0.9709
Baby, I'm just gonna shake, shake, shake, shake, shake: ==> -0.6705
I shake it off, I shake it off: ==> -0.34
Heart-breakers gonna break, break, break, break, break: ==> 0.0
And the fakers gonna fake, fake, fake, fake, fake: ==> -0.9382
Baby, I'm just gonna shake, shake, shake, shake, shake: ==> -0.6705
I shake it off, I shake it off: ==> -0.34
I never miss a beat: ==> 0.1139
I'm lightning on m

According to vader_lexicon "Shake It Off" is apparently a 
slightly negative jam with an aggregated -0.19 score below neutral... LOL
